src/ripple/unl/tests/BasicNetwork.h

```diff
@@ -57,132 +60,176 @@ class BasicNetwork
         typename clock_type::time_point;
 
 private:
-    struct invoke
-    {
-        virtual ~invoke() = default;
-        virtual void operator()(Peer& from) const = 0;
-    };
-
-    template <class Message>
-    class invoke_impl;
+    struct by_to_tag {};
+    struct by_from_tag {};
+    struct by_when_tag {};
+
+    using by_to_hook =
+        boost::intrusive::list_base_hook<
+            boost::intrusive::link_mode<
+                boost::intrusive::normal_link>,
+            boost::intrusive::tag<by_to_tag>>;
+
+    using by_from_hook =
+        boost::intrusive::list_base_hook<
+            boost::intrusive::link_mode<
+                boost::intrusive::normal_link>,
+            boost::intrusive::tag<by_from_tag>>;
+
+    using by_when_hook =
+        boost::intrusive::set_base_hook<
+            boost::intrusive::link_mode<
+                boost::intrusive::normal_link>>;
 
     struct msg
+        : by_to_hook, by_from_hook, by_when_hook
     {
-        msg (Peer& from_, Peer& to_, time_point when_,
-                std::unique_ptr<invoke> op_)
-            : to(&to_), from(&from_),
-                when(when_), op(std::move(op_))
-        {
-        }
-
         Peer* to;
         Peer* from;
         time_point when;
-        std::unique_ptr<invoke> mutable op;
+
+        virtual ~msg() = default;
+        virtual void operator()() const = 0;
+
+        msg (Peer* from_, Peer* to_, time_point when_)
+            : to(to_), from(from_), when(when_)
+        {
+        }
+
+        bool
+        operator< (msg const& other) const
+        {
+            return when < other.when;
+        }
     };
 
-    struct link_type
+    template <class Handler>
+    class msg_impl : public msg
     {
-        bool inbound;
-        duration delay;
+    private:
+        Handler const h_;
 
-        link_type (bool inbound_,
-                duration delay_)
-            : inbound (inbound_)
-            , delay (delay_)
+    public:
+        msg_impl (Peer* from_, Peer* to_,
+                time_point when_, Handler&& h)
```

Comment:
Just a note that `Handler` here is an rvalue reference (as opposed to a forwarding reference) because it is not a deduced type. If `Handler` is cheap to move, I'd combine this constructor with the one below by moving from a value type (`Handler` rather than `Handler&&`), but we've had this discussion before and I certainly don't object to what's here.